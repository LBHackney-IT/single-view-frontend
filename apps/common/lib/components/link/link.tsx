import React, { forwardRef } from "react";

import cn from "classnames";

import { widthOverrides } from "../../utils";

import type * as Polymorphic from "@radix-ui/react-polymorphic";
import "./styles.scss";

export interface LinkProps {
  variant?: "link" | "danger" | "text-colour" | "muted" | "back-link" | "native";
  isExternal?: boolean;
  override?: number;
}

export type LinkComponent = Polymorphic.ForwardRefComponent<"a", LinkProps>;

export const Link: LinkComponent = forwardRef(function Link(
  {
    as: LinkComp = "a",
    variant = "link",
    isExternal = false,
    className,
    rel,
    target,
    override,
    ...props
  },
  ref,
) {
  const linkClasses = cn(
    variant !== "native" && {
      "govuk-link lbh-link": variant !== "back-link",
      "govuk-back-link lbh-back-link": variant === "back-link",
      [`lbh-link--${variant}`]: variant !== "link" && variant !== "back-link",
      "lbh-link--no-visited-state": !isExternal,
    },
    widthOverrides(override),
    className,
  );
  return (
    // eslint-disable-next-line react/jsx-no-target-blank
    <LinkComp
      ref={ref}
      className={linkClasses}
      rel={isExternal ? "noopener noreferrer" : rel}
      target={isExternal ? "_blank" : target}
      {...props}
    />
  );
});
