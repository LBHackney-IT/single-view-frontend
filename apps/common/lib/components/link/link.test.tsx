import React, { useEffect } from "react";

import { render, testA11y } from "@hackney/mtfh-test-utils";

import { Link } from "./link";

test("it renders correctly", async () => {
  const { container } = render(<Link href="http://localhost">Test Link</Link>);
  expect(container).toMatchSnapshot();
  await testA11y(container);
});

test("it renders a variant", () => {
  const { container } = render(
    <Link href="http://localhost" variant="muted">
      Test Link
    </Link>,
  );
  expect(container).toMatchSnapshot();
});

test("it renders a back link", () => {
  const { container } = render(
    <Link href="http://localhost" variant="back-link">
      Test Link
    </Link>,
  );
  expect(container).toMatchSnapshot();
});

test("it adds the correct attributes for external links", () => {
  const { container } = render(
    <Link href="http://localhost" isExternal>
      Test Link
    </Link>,
  );
  expect(container).toMatchSnapshot();
});

test("it accepts a ref", () => {
  const callback = jest.fn();
  const Comp = () => {
    const ref = React.createRef<HTMLAnchorElement>();

    useEffect(() => {
      if (ref.current) {
        callback(ref.current);
      }
    }, [ref]);
    return (
      <Link ref={ref} href="http://localhost">
        Test Link
      </Link>
    );
  };
  render(<Comp />);
  expect(callback).toBeCalledTimes(1);
});
