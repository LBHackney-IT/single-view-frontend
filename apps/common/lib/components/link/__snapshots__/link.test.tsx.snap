// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it adds the correct attributes for external links 1`] = `
<div>
  <a
    class="govuk-link lbh-link"
    href="http://localhost"
    rel="noopener noreferrer"
    target="_blank"
  >
    Test Link
  </a>
</div>
`;

exports[`it renders a back link 1`] = `
<div>
  <a
    class="govuk-back-link lbh-back-link lbh-link--no-visited-state"
    href="http://localhost"
  >
    Test Link
  </a>
</div>
`;

exports[`it renders a variant 1`] = `
<div>
  <a
    class="govuk-link lbh-link lbh-link--muted lbh-link--no-visited-state"
    href="http://localhost"
  >
    Test Link
  </a>
</div>
`;

exports[`it renders correctly 1`] = `
<div>
  <a
    class="govuk-link lbh-link lbh-link--no-visited-state"
    href="http://localhost"
  >
    Test Link
  </a>
</div>
`;
